torchrl/data/replay_buffers/rb_prototype.py

```diff
@@ -155,6 +156,17 @@ def sample(self, batch_size: int) -> Tuple[Any, dict]:
 
         return ret
 
+    def mark_update(self, index) -> None:
+        """Marks a given storage index as having changed.
+
+        Derived classes can deal with this however appropriate,
+        forwarding this call to whichever parts are needed.
+
+        Args:
+            index: The modified index from storage.
+        """
+        return self.update_priority(index, self._sampler.default_priority)
```

Comment:
That's cool
In the future we will probably want the user to update the priority of the sampler directly, without passing through the replay buffer (otherwise the replay buffer will need to implement custom update methods for all the modules e put in it).

One fix for now could be that the sampler has a `mark_update` method that falls back on `update_priority` for prioritized sampling. If sampling is uniform, `mark_update` is a no-op.
Like this we could remove the `ReplayBuffer.update_priority` method altogether (in a future PR).

Would that make sense?

Reply:
It makes sense, yes.

I debated with myself for a while if we should attach samplers to storage or if we should attach RBs to storage, and decided for the latter because it is more extensible: It's possible that we need components other than the sampler to be aware of changes, and it's the RB's duty to coordinate the different moving parts.

The counterpoint to this would be YAGNI. It's simpler if we attach the samplers directly. I can definitely make that change.

By the way, another advantage of having the `mark_update` method instead of directly calling `update_priority` is that samplers can implement the update whenever they feel like it. We could, for example, simply store the modified indexes inside the sampler until the next call to sample, then update everything lazily.

Reply:
> I debated with myself for a while if we should attach samplers to storage or if we should attach RBs to storage, and decided for the latter because it is more extensible: It's possible that we need components other than the sampler to be aware of changes, and it's the RB's duty to coordinate the different moving parts.

I fully agree. I still think we should attach the RB to the storage, not all other children (the parent is the minimal sufficient information).

My point is mostly about generality of the name `update_priority` in the replay buffer. I would rather name that `mark_update`:

```python
class ReplayBuffer:
    def mark_update(self, ...):
        self.sampler.mark_update(...)

class PrioritizedSampler:
    def mark_update(self, ...)
        self.update_priority(...)
   def update_priority(self, ...):
        foo(...)
```

> By the way, another advantage of having the mark_update method instead of directly calling update_priority is that samplers can implement the update whenever they feel like it. We could, for example, simply store the modified indexes inside the sampler until the next call to sample, then update everything lazily.

Yes that would be cool! That would definitely be an advantage (we could also update things in batch when needed, which I would expect to cut the compute time of those ops by a bit)

Reply:
Ah yes. Just made that change.